changelog.d/12319.bugfix

```diff
@@ -0,0 +1 @@
+Fix bug with incremental sync missing events when rejoining/backfilling. Contributed by Nick @ Beeper.
```

Comment:
For the reviewer: The CI failures described in https://github.com/matrix-org/synapse/pull/12319#discussion_r839267470 are fixed by https://github.com/matrix-org/synapse/pull/12370

The first thing to decide is whether we like the fix from this PR. If we like the fix there, then we also need to consider the fix in https://github.com/matrix-org/synapse/pull/12370 to fix the `/messages` bug that was uncovered in this PR.

Reply:
@Fizzadar https://github.com/matrix-org/synapse/pull/12370 is now merged. Merge `develop`  into this PR to see the tests passing again ✅

Reply:
Pushed, and complement passed OK! Not had a chance to look at the sytest failure yet!